fuzzers/main.py

```diff
@@ -0,0 +1,450 @@
+        while proc.poll() == None and not exit_event.is_set():
+            time.sleep(0.5)
+            new_time = time.monotonic()
+            fuzz_time.inc(new_time - last_time)
+            last_time = new_time
+        print(f"Fuzzer running {t} has stopped")
+
+    # Remove this fuzzer from the fuzzer list
+    FUZZERS.remove(fuzzer)
+
+    # If a crash was found, report
+    if proc.poll() != None:
+        FUZZ_CRASHES.labels(branch['name'], t['crate'], t['runner'], t['flags']).inc()
+
+    # Stop the inotify threads and the process, and make sure to wait until they are stopped
+    stop_inotify_event.set()
+    if proc.poll() == None:
+        signal_fuzzer(signal.SIGTERM, proc)
+    time.sleep(5)
+    if proc.poll() == None:
+        signal_fuzzer(signal.SIGKILL, proc)
+
+    # Finally, upload the log files
+    bucket.blob(f'logs/{log_path}').upload_from_filename(log_filename)
+
+def start_a_fuzzer(branch: str, cfg: ConfigType, gcs: google.cloud.storage.Client) -> None:
+    """Start a fuzzer that will randomly run one of the fuzz targets from `cfg` forever"""
+    def run_it(cfg: ConfigType, gcs: google.cloud.storage.Client) -> None:
+        exit_event = threading.Event()
+        while not exit_event.is_set():
+            target = random.choices(cfg['target'], [t['weight'] for t in cfg['target']])[0]
+            run_fuzzer(branch, target, gcs, exit_event)
+    threading.Thread(daemon = True, target = run_it, args = (cfg, gcs)).start()
+
+def signal_fuzzer(signal: int, proc: subprocess.Popen) -> None:
+    """
+    Kill fuzzer `proc` (with all its process group as a fuzzer can spawn sub-processes) with
+    `signal`
+    """
+
+    print(f"Killing fuzzer {proc.pid} with signal {signal}")
+    os.killpg(os.getpgid(proc.pid), signal)
+
+def signal_fuzzers(signal: int) -> None:
+    """
+    Kill all fuzzers (with all their process group as a fuzzer can spawn sub-processes) with
+    `signal`
+    """
+
+    for f in FUZZERS:
+        signal_fuzzer(signal, f.proc)
+
+def stop_fuzzers() -> None:
+    """Stop all the fuzzers, cleaning everything up"""
+    global FUZZERS
+    signal_fuzzers(signal.SIGTERM)
+    for f in FUZZERS:
+        f.exit_event.set()
+    time.sleep(5)
+    signal_fuzzers(signal.SIGKILL)
+    FUZZERS = []
+
+def listen_for_commands(gcs: google.cloud.storage.Client) -> None:
+    """
+    Spawn an HTTP server to remote control the fuzzers
+
+    `/pause` and `/resume` respectively sigstop and sigcont the fuzzers
+    `/restart` forces a fuzzer restar
+    `/exit` stops the whole process immediately (without cleanup)
+    """
+    class HTTPHandler(BaseHTTPRequestHandler):
+        def do_GET(self) -> None:
+            global FUZZERS
+            if self.path == '/pause':
+                signal_fuzzers(signal.SIGSTOP)
+                self.send_response(200)
+            elif self.path == '/resume':
+                signal_fuzzers(signal.SIGCONT)
+                self.send_response(200)
```

Comment:
So there is one race condition with this at the moment. If we `/pause` while fuzzers aren’t running yet (e.g. when fresh master branch from the repository is fetched), the fuzzers will start even though `/resume` wasn’t sent. This is probably not critical at the moment and we can start testing even with this issue, but this should be addressed.